spec-0007/index.md

```diff
@@ -97,126 +100,169 @@ The [deprecation strategy](https://github.com/scientific-python/specs/pull/180#i
-        # 1. If np.random.seed(None) was called (or never called), then we can
-        #    just use the default_rng (the result is random anyway).
-        # 2. If it was called, we must return the global random state object
-        #    and warn about future ignoring of seed!
-        if np.random.mtrand._rand._bit_generator._seed_seq is not None:
-            # The user did not seed, so no need to warn.
-            return np.random.default_rng()
-        warnings.warn(
-            "The NumPy global rng was seeded in call to np.random.seed() "
-            "in the future this function will ignore this seed and return "
-            "random values as if a new `np.random.default_rng()` was created.",
-            FutureWarning, stacklevel=5)
-        return np.random.mtrand._rand
-    if seed is None or seed is np.random:
-        return np.random.mtrand._rand
-    if isinstance(seed, (numbers.Integral, np.integer)):
-        return np.random.RandomState(seed)
-    if isinstance(seed, (np.random.RandomState, np.random.Generator)):
-        return seed
-
-    raise ValueError(f"'{seed}' cannot be used to seed a numpy.random.RandomState"
-                     " instance")
-
-
-def _prepare_rng(old_name, dep_version=None):
+import numpy as np
+import functools
+import warnings
+
+def _transition_to_rng(old_name, position_num=None, dep_version=None):
+    """ Example decorator to transition from old PRNG usage to new `rng` behavior
+
+    Suppose the decorator is applied to a function that used to accept parameter
+    `old_name='random_state'` either by keyword or as a positional argument at
+    `position_name=1`. At the time of application, the name of the argument in the
+    function signature is manually changed to the new name, `rng`.
+
+    - If the function is called with both `random_state` and `rng`, the decorator
+      raises an error.
+    - If `random_state` is provided as a keyword argument, the decorator passes
+      `random_state` to the function's `rng` argument as a keyword, but the decorator
+      will emit a `DeprecationWarning` about the deprecation of keyword `random_state`.
+    - If `random_state` is provided as a positional argument, the decorator passes
+      `random_state` to the function's `rng` argument by position, but the decorator
+      will emit a `FutureWarning` about the changing behavior of the argument.
+    - If `rng` is provided as a keyword argument, the decorator validates `rng` using
+      `numpy.random.default_rng` before passing it to the function.
+    - If neither `random_state` nor `rng` is provided, the decorator checks whether
+      `np.random.seed` has been used to set the global seed. If so, it emits a
+      `FutureWarning` against use of `numpy.random.seed`. Either way, the decorator
+      calls the function without explicitly passing the `rng` argument.
+
+    To avoid warnings, a user must pass `rng` as a keyword or pass a `Generator`
+    object or `None` by position.
```

Comment:
We discussed, but I still wonder if we aren't being too lenient. If the library lets this go past, the *same code* will produce very different results in old and new versions?

If they error on positional None, then at least they *know* when None is sent to mean "use system entropy" vs "use global seed state".

Reply:
I realized this was no good and took out this behavior a while ago. It will always warn with positional input now. I'll change this part of the text.